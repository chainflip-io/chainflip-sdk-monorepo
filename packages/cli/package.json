{
  "name": "@chainflip-io/cli",
  "version": "0.1.4",
  "description": "",
  "main": "./dist/lib.mjs",
  "types": "./dist/lib.d.mts",
  "bin": {
    "chainflip": "./dist/cli.mjs"
  },
  "exports": {
    ".": {
      "import": "./dist/lib.mjs",
      "types": "./dist/lib.d.mts"
    }
  },
  "typesVersions": {
    "*": {
      ".": [
        "dist/lib.d.mts"
      ]
    }
  },
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "build": "rm -rf dist && tsup",
    "test": "jest",
    "test:ci": "CI=1 jest --forceExit",
    "typecheck": "tsc --noEmit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "ethers": "6.8.0",
    "yargs": "^17.7.2"
  },
  "devDependencies": {
    "@types/yargs": "^17.0.26",
    "tsup": "^7.2.0",
    "tsx": "^3.13.0"
  },
  "repository": "https://github.com/chainflip-io/chainflip-sdk-monorepo.git",
  "publishConfig": {
    "registry": "https://npm.pkg.github.com"
  }
}
