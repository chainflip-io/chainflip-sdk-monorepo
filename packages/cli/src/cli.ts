import yargs from 'yargs/yargs';
import cliEncodeVaultSwapData, {
  yargsOptions as cliEncodeVaultSwapDataOptions,
} from './commands/cliEncodeVaultSwapData';
import cliFundStateChainAccount, {
  yargsOptions as cliFundStateChainAccountOptions,
} from './commands/cliFundStateChainAccount';
import cliRequestSwapDepositAddress, {
  yargsOptions as cliRequestSwapDepositAddressOptions,
} from './commands/cliRequestSwapDepositAddress';

export default async function cli(args: string[]) {
  return yargs(args)
    .scriptName('chainflip-cli')
    .usage('$0 <cmd> [args]')
    .command(
      'fund-state-chain-account',
      '',
      cliFundStateChainAccountOptions,
      cliFundStateChainAccount,
    )
    .command(
      'request-swap-deposit-address',
      '',
      cliRequestSwapDepositAddressOptions,
      cliRequestSwapDepositAddress,
    )
    .command('get-vault-swap-data', '', cliEncodeVaultSwapDataOptions, cliEncodeVaultSwapData)
    .wrap(0)
    .strict()
    .help()
    .parse();
}
