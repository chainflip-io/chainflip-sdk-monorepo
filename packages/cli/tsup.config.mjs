/* eslint-disable import/no-extraneous-dependencies */
import { defineConfig } from 'tsup';

export default defineConfig({
  treeshake: true,
  minify: false,
  dts: true,
  format: ['cjs', 'esm'],
  entry: {
    lib: 'src/lib/index.ts',
    cli: 'src/main.ts',
  },
  splitting: true,
});
