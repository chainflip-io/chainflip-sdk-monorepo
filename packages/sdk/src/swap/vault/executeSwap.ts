import assert from 'assert';
import { ContractReceipt, Signer } from 'ethers';
import { z } from 'zod';
import {
  SISYPHOS_FLIP_CONTRACT_ADDRESS,
  GOERLI_USDC_CONTRACT_ADDRESS,
  SISYPHOS_VAULT_CONTRACT_ADDRESS,
} from '@/shared/addresses';
import {
  ChainflipNetwork,
  SupportedAsset,
  chainflipNetwork,
} from '@/shared/enums';
import {
  Vault,
  Vault__factory,
  ERC20__factory,
} from '../../../types/ethers-contracts';
import { ChainId } from '../sdk';
import { isTestnet } from '../utils';
import {
  ExecuteSwapParams,
  NativeSwapParams,
  TokenSwapParams,
  executeSwapParamsSchema,
} from './validators';

// !!!!! IMPORTANT !!!!!
// Do not change these indices.
const chainMap: Record<ChainId, number> = {
  [ChainId.Ethereum]: 1,
  [ChainId.Polkadot]: 2,
  [ChainId.Bitcoin]: 3,
};

// !!!!!! IMPORTANT !!!!!!
// Do not change these indices.
const assetMap: Record<SupportedAsset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  ETH: 1,
  FLIP: 2,
  USDC: 3,
  DOT: 4,
  BTC: 5,
};

const swapNative = async (
  vault: Vault,
  { destChainId, destTokenSymbol, destAddress, amount }: NativeSwapParams,
): Promise<ContractReceipt> => {
  const transaction = await vault.xSwapNative(
    chainMap[destChainId],
    destAddress,
    assetMap[destTokenSymbol],
    [],
    { value: amount },
  );

  const receipt = await transaction.wait(1);

  assert(receipt.status !== 0, 'Transaction failed');

  return receipt;
};

const getTokenContractAddress = (
  asset: SupportedAsset,
  cfNetwork: ChainflipNetwork,
): string => {
  assert(isTestnet(cfNetwork), 'Only testnets are supported for now');

  if (asset === 'FLIP' && cfNetwork === 'sisyphos') {
    return SISYPHOS_FLIP_CONTRACT_ADDRESS;
  }

  assert(asset === 'USDC');

  return GOERLI_USDC_CONTRACT_ADDRESS;
};

const swapToken = async (
  vault: Vault,
  params: TokenSwapParams,
  { signer, ...opts }: ExecuteSwapOptions,
): Promise<ContractReceipt> => {
  const erc20Address =
    opts.cfNetwork === 'localnet'
      ? opts.srcTokenContractAddress
      : getTokenContractAddress(params.srcTokenSymbol, opts.cfNetwork);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');
  const erc20 = ERC20__factory.connect(erc20Address, signer);
  const signerAddress = await signer.getAddress();
  const allowance = await erc20.allowance(signerAddress, vault.address);

  if (allowance.lt(params.amount)) {
    const approval = await erc20.approve(vault.address, params.amount);
    const approvalReceipt = await approval.wait(1);
    assert(approvalReceipt.status !== 0, 'Approval failed');
  }

  const transaction = await vault.xSwapToken(
    chainMap[params.destChainId],
    params.destAddress,
    assetMap[params.destTokenSymbol],
    erc20Address,
    params.amount,
    [],
  );

  const receipt = await transaction.wait(1);

  assert(receipt.status !== 0, 'Transaction failed');

  return receipt;
};

const isTokenSwap = (params: ExecuteSwapParams): params is TokenSwapParams =>
  'srcTokenSymbol' in params;

const executeSwapOptionsSchema = z.intersection(
  z.object({ signer: z.instanceof(Signer) }),
  z.union([
    z.object({ cfNetwork: chainflipNetwork }),
    z.object({
      cfNetwork: z.literal('localnet'),
      vaultContractAddress: z.string(),
      srcTokenContractAddress: z.string().optional(),
    }),
  ]),
);

export type ExecuteSwapOptions = z.infer<typeof executeSwapOptionsSchema>;

const executeSwap = async (
  params: ExecuteSwapParams,
  options: ExecuteSwapOptions,
): Promise<ContractReceipt> => {
  executeSwapParamsSchema.parse(params);
  const opts = executeSwapOptionsSchema.parse(options);

  let vaultContractAddress: string | undefined;
  if (opts.cfNetwork === 'localnet') {
    vaultContractAddress = opts.vaultContractAddress;
  } else if (opts.cfNetwork === 'sisyphos') {
    vaultContractAddress = SISYPHOS_VAULT_CONTRACT_ADDRESS;
  }

  assert(
    vaultContractAddress !== undefined,
    'Missing vault contract address or network unsupported',
  );

  const vault = Vault__factory.connect(vaultContractAddress, opts.signer);

  if (isTokenSwap(params)) return swapToken(vault, params, opts);
  return swapNative(vault, params);
};

export default executeSwap;
