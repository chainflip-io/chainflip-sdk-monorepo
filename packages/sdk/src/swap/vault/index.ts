export { default as executeSwap } from './executeSwap';
export type { ExecuteSwapParams } from './validators';
