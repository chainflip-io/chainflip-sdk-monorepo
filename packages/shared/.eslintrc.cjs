// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require('path');

module.exports = {
  extends: '../../.eslintrc.json',
  plugins: ['eslint-plugin-n'],
  rules: {
    'n/no-process-env': ['error'],
    'import/no-extraneous-dependencies': [
      'error',
      {
        packageDir: [__dirname, path.join(__dirname, '..', '..')],
      },
    ],
  },
  overrides: [
    {
      files: ['*.test.ts', '*.mjs', '*.js', '*.cjs'],
      rules: {
        'n/no-process-env': ['off'],
      },
    },
  ],
};
