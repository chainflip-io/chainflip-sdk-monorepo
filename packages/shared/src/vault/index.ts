export { default as executeSwap, type ExecuteSwapOptions } from './executeSwap';
export type { ExecuteSwapParams } from './validators';
export * from './approval';
