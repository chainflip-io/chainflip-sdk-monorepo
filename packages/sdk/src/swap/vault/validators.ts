import { z } from 'zod';
import { supportedAsset } from '@/shared/enums';
import {
  btcAddress,
  dotAddress,
  ethereumAddress,
  numericString,
} from '@/shared/parsers';
import { ChainId } from '../sdk';

const base = z.object({ amount: numericString });

const ethereumBase = base.extend({
  destChainId: z.literal(ChainId.Ethereum),
  destAddress: ethereumAddress,
});

const polkadotBase = base.extend({
  destChainId: z.literal(ChainId.Polkadot),
  destAddress: dotAddress,
});

const bitcoinBase = base.extend({
  destChainId: z.literal(ChainId.Bitcoin),
  destAddress: btcAddress,
});

const erc20 = z.union([
  z.literal(supportedAsset.enum.FLIP),
  z.literal(supportedAsset.enum.USDC),
]);

const ethereumNative = ethereumBase
  .extend({
    destTokenSymbol: erc20,
  })
  .strict();

const polkadotNative = polkadotBase
  .extend({
    destTokenSymbol: z.literal(supportedAsset.enum.DOT),
    srcTokenSymbol: z.undefined(),
  })
  .strict();

const bitcoinNative = bitcoinBase
  .extend({
    destTokenSymbol: z.literal(supportedAsset.enum.BTC),
    srcTokenSymbol: z.undefined(),
  })
  .strict();

const nativeSwapParamsSchema = z.union([
  ethereumNative,
  polkadotNative,
  bitcoinNative,
]);

export type NativeSwapParams = z.infer<typeof nativeSwapParamsSchema>;

const tokenSwapParamsSchema = z.union([
  ethereumBase.extend({
    srcTokenSymbol: z.literal(supportedAsset.enum.FLIP),
    destTokenSymbol: z.union([
      z.literal(supportedAsset.enum.USDC),
      z.literal(supportedAsset.enum.ETH),
    ]),
  }),
  ethereumBase.extend({
    srcTokenSymbol: z.literal(supportedAsset.enum.USDC),
    destTokenSymbol: z.union([
      z.literal(supportedAsset.enum.FLIP),
      z.literal(supportedAsset.enum.ETH),
    ]),
  }),
  polkadotNative.extend({ srcTokenSymbol: erc20 }),
  bitcoinNative.extend({ srcTokenSymbol: erc20 }),
]);

export type TokenSwapParams = z.infer<typeof tokenSwapParamsSchema>;

export const executeSwapParamsSchema = z.union([
  nativeSwapParamsSchema,
  tokenSwapParamsSchema,
]);

export type ExecuteSwapParams = z.infer<typeof executeSwapParamsSchema>;
