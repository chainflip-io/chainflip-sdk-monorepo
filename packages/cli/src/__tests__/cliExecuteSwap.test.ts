import { executeSwap } from '@/shared/vault';
import cli from '../cli';

jest.mock('ethers');
jest.mock('@/shared/vault', () => ({
  executeSwap: jest
    .fn()
    .mockResolvedValue({ status: 1, hash: 'example-tx-hash' }),
}));

const localnet = `swap
  --wallet-private-key 0x2
  --chainflip-network localnet
  --src-asset ETH
  --dest-asset USDC
  --amount 1000000000
  --dest-address 0x0
  --src-token-contract-address 0x0
  --vault-contract-address 0x0
  --eth-network test`;

const localnetWithCcm = `swap
  --wallet-private-key 0x2
  --chainflip-network localnet
  --src-asset ETH
  --dest-asset USDC
  --amount 1000000000
  --dest-address 0x0
  --src-token-contract-address 0x0
  --vault-contract-address 0x0
  --eth-network test
  --message=0xdeadc0de
  --gas-budget=500000`;

describe('cli', () => {
  it.each([localnet, localnet.replace('localnet', 'sisyphos')])(
    'calls the correct handler with the proper arguments (swap)',
    async (args) => {
      const logSpy = jest.spyOn(global.console, 'log').mockImplementation();
      await cli(args.split(/\s+/));

      expect(executeSwap).toHaveBeenCalledTimes(1);
      expect(jest.mocked(executeSwap).mock.calls[0][0]).toHaveProperty(
        'ccmMetadata',
        undefined,
      );
      expect(jest.mocked(executeSwap).mock.calls).toMatchSnapshot();
      expect(logSpy).toHaveBeenCalledWith(
        expect.stringContaining('Transaction hash: example-tx-hash'),
      );
    },
  );

  it.each([localnetWithCcm, localnetWithCcm.replace('localnet', 'sisyphos')])(
    'calls the correct handler with the proper arguments (call)',
    async (args) => {
      const logSpy = jest.spyOn(global.console, 'log').mockImplementation();
      await cli(args.split(/\s+/));

      expect(executeSwap).toHaveBeenCalledTimes(1);
      expect(jest.mocked(executeSwap).mock.calls[0][0]).toHaveProperty(
        'ccmMetadata.message',
        '0xdeadc0de',
      );
      expect(jest.mocked(executeSwap).mock.calls[0][0]).toHaveProperty(
        'ccmMetadata.gasBudget',
        '500000',
      );
      expect(jest.mocked(executeSwap).mock.calls).toMatchSnapshot();
      expect(logSpy).toHaveBeenCalledWith(
        expect.stringContaining('Transaction hash: example-tx-hash'),
      );
    },
  );
});
