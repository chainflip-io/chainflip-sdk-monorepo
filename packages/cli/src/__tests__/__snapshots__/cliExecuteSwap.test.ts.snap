// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli calls the correct handler with the proper arguments (call) 1`] = `
[
  [
    {
      "amount": "1000000000",
      "ccmMetadata": {
        "gasBudget": "500000",
        "message": "0xdeadc0de",
      },
      "destAddress": "0x0",
      "destAsset": "USDC",
      "destChain": "Ethereum",
      "srcAsset": "ETH",
      "srcChain": "Ethereum",
    },
    {
      "network": "localnet",
      "signer": undefined,
      "srcTokenContractAddress": "0x0",
      "vaultContractAddress": "0x0",
    },
  ],
]
`;

exports[`cli calls the correct handler with the proper arguments (call) 2`] = `
[
  [
    {
      "amount": "1000000000",
      "ccmMetadata": {
        "gasBudget": "500000",
        "message": "0xdeadc0de",
      },
      "destAddress": "0x0",
      "destAsset": "USDC",
      "destChain": "Ethereum",
      "srcAsset": "ETH",
      "srcChain": "Ethereum",
    },
    {
      "network": "sisyphos",
      "signer": undefined,
    },
  ],
]
`;

exports[`cli calls the correct handler with the proper arguments (swap) 1`] = `
[
  [
    {
      "amount": "1000000000",
      "ccmMetadata": undefined,
      "destAddress": "0x0",
      "destAsset": "USDC",
      "destChain": "Ethereum",
      "srcAsset": "ETH",
      "srcChain": "Ethereum",
    },
    {
      "network": "localnet",
      "signer": undefined,
      "srcTokenContractAddress": "0x0",
      "vaultContractAddress": "0x0",
    },
  ],
]
`;

exports[`cli calls the correct handler with the proper arguments (swap) 2`] = `
[
  [
    {
      "amount": "1000000000",
      "ccmMetadata": undefined,
      "destAddress": "0x0",
      "destAsset": "USDC",
      "destChain": "Ethereum",
      "srcAsset": "ETH",
      "srcChain": "Ethereum",
    },
    {
      "network": "sisyphos",
      "signer": undefined,
    },
  ],
]
`;
